Generate pseudo-class serialization and pseudo-element aliasing

nt1m opened the pull request:
#### 9bfed8cf528434bbb0eb6ad5e5fe716cfbcbb533
<pre>
Generate pseudo-class serialization and pseudo-element aliasing
<a href="https://bugs.webkit.org/show_bug.cgi?id=266978">https://bugs.webkit.org/show_bug.cgi?id=266978</a>
<a href="https://rdar.apple.com/120359881">rdar://120359881</a>

Reviewed by Darin Adler.

- Pseudo-class serialization is mechanical, generate most of it. We don&apos;t re-use the same approach as pseudo-elements for serialization
since that initializes CSSSelector rare data, and also because each possible value has a 1:1 mapping to the PseudoClass enum, unlike for PseudoElement.
- Generate code for aliasing at parse-time for `PseudoElement::WebKitCustomLegacyPrefixed`

This commit gets us closer to having CSSPseudoSelectors.json becoming a single source of truth.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/parser/CSSParserSelector.cpp:
(WebCore::CSSParserSelector::parsePseudoElementSelector):
* Source/WebCore/css/process-css-pseudo-selectors.py:

Canonical link: <a href="https://commits.webkit.org/272583@main">https://commits.webkit.org/272583@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb5bf9569a65492499188bc2e359e65105ca5f06

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32213 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10947 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34027 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34773 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29192 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33063 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13300 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8157 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28761 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32631 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9228 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28799 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8037 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8188 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28717 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36118 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29288 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/29166 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34315 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8313 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/6264 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32178 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9954 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/7522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8945 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4172 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8852 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -119,6 +119,9 @@ struct PossiblyQuotedIdentifier {
         static std::optional<PseudoElement> parsePseudoElement(StringView, const CSSSelectorParserContext&);
         static std::optional<PseudoId> parseStandalonePseudoElement(StringView, const CSSSelectorParserContext&);
 
+        static ASCIILiteral selectorTextForPseudoClass(CSSSelector::PseudoClass type);
```
I don't think this argument needs a name.

darinadler commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -119,6 +119,9 @@ struct PossiblyQuotedIdentifier {
         static std::optional<PseudoId> parseStandalonePseudoElement(StringView, const CSSSelectorParserContext&);
 
+        static ASCIILiteral selectorTextForPseudoClass(CSSSelector::PseudoClass type);
+        static ASCIILiteral nameForShadowPseudoElementLegacyAlias(StringView type);
```
Might be clearer without an argument name.

nt1m committed:
Generate pseudo-class serialization and pseudo-element aliasing
https://bugs.webkit.org/show_bug.cgi?id=266978
rdar://120359881

Reviewed by Darin Adler.

- Pseudo-class serialization is mechanical, generate most of it. We don't re-use the same approach as pseudo-elements for serialization
since that initializes CSSSelector rare data, and also because each possible value has a 1:1 mapping to the PseudoClass enum, unlike for PseudoElement.
- Generate code for aliasing at parse-time for `PseudoElement::WebKitCustomLegacyPrefixed`

This commit gets us closer to having CSSPseudoSelectors.json becoming a single source of truth.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/parser/CSSParserSelector.cpp:
(WebCore::CSSParserSelector::parsePseudoElementSelector):
* Source/WebCore/css/process-css-pseudo-selectors.py:

Canonical link: https://commits.webkit.org/272583@main
